app/client/src/components/shared/AddSaveDataWidget.SavePanel.tsx

```diff
@@ -0,0 +1,722 @@
+  'cyanLayer',
+];
+
+const layerTypesToIgnore = ['wcs', 'wfs'];
+
+const layersRequireFeatureService = [
+  'actionsLayer',
+  'boundariesLayer',
+  'cyanLayer',
+  'dischargersLayer',
+  'issuesLayer',
+  'monitoringLocationsLayer',
+  'providersLayer',
+  'searchIconLayer',
+  'usgsStreamgagesLayer',
+  'waterbodyLayer',
+  'upstreamWatershed',
+];
+
+const layersMayBeFiltered = [
+  'dischargersLayer',
+  'issuesLayer',
+  'monitoringLocationsLayer',
+];
+
+const layersMayNotHaveLoaded = ['cyanLayer', 'upstreamWatershed'];
+
+const layerFilterOptions = [
+  { value: 'All', label: 'All' },
+  { value: 'Free', label: 'Free' },
+];
+
+// Performs a deep comparison of 2 objects. Returns true if they are equal
+// and false otherwise.
+function deepEqual(object1: any, object2: any) {
+  const keys1 = Object.keys(object1);
+  const keys2 = Object.keys(object2);
+  if (keys1.length !== keys2.length) {
+    return false;
+  }
+  for (const key of keys1) {
+    const val1 = object1[key];
+    const val2 = object2[key];
+    const areObjects = isObject(val1) && isObject(val2);
+    if (
+      (areObjects && !deepEqual(val1, val2)) ||
+      (!areObjects && val1 !== val2)
+    ) {
+      return false;
+    }
+  }
+  return true;
+}
+function isObject(object: any) {
+  return object != null && typeof object === 'object';
+}
+
+const modifiedErrorBoxStyles = css`
+  ${errorBoxStyles};
+  text-align: center;
+  margin-bottom: 0.625rem;
+`;
+
+const modifiedSuccessBoxStyles = css`
+  ${successBoxStyles};
+  text-align: center;
+  margin-bottom: 0.625rem;
+`;
+
+const listContainerStyles = css`
+  margin: 0.625rem 0;
+`;
+
+const switchStyles = css`
+  display: flex;
+  align-items: center;
+  margin-bottom: 0.625rem;
+  gap: 0.5rem;
+`;
+
+const submitSectionStyles = css`
+  margin-top: 1.5rem;
+  margin-bottom: 0.625rem;
+`;
+
+const buttonContainerStyles = css`
+  display: flex;
+  justify-content: flex-end;
+  align-items: center;
+`;
+
+const addButtonStyles = css`
+  margin: 0;
+  min-width: 50%;
+  font-weight: normal;
+  font-size: 0.75rem;
+`;
+
+const sharedInputStyles = `
+  border-width: 1px;
+  border-style: solid;
+  border-radius: 4px;
+  border-color: hsl(0, 0%, 80%);
+  margin-bottom: 10px;
+  padding: 2px 8px;
+  width: 100%;
+`;
+
+const saveAsInputStyles = css`
+  ${sharedInputStyles}
+  height: 36px;
+`;
+
+const descriptionInputStyles = css`
+  ${sharedInputStyles}
+  min-height: 36px;
+  height: 72px;
+  resize: vertical;
+`;
+
+function errorMessage(text: string) {
+  return (
+    <div css={modifiedErrorBoxStyles}>
+      <p>{text}</p>
+    </div>
+  );
+}
+
+type Props = {
+  visible: boolean;
+};
+
+function SavePanel({ visible }: Props) {
+  const {
+    saveAsName,
+    setSaveAsName,
+    saveDescription,
+    setSaveDescription,
+    saveLayersList,
+    setSaveLayersList,
+    widgetLayers,
+  } = useAddSaveDataWidgetState();
+  const mapView = useContext(LocationSearchContext).mapView as __esri.MapView;
```

Comment:
This type cast is iffy, because `mapView` could technically be `null`, but I think in this case it's fine, since there's no way to open the widget while the map is loading.

Reply:
I didn't really like doing this. I just did it to get some kind of type, even though the LocationSearchContext isn't typescript ready yet. Once we get the LocationSearchContext updated for TypeScript we can get rid of this code.

I will go ahead and change this to `const mapView = useContext(LocationSearchContext).mapView as __esri.MapView | null;` so the type is atleast accurate.